src/Estimators/SpinDensityNew.cpp

```diff
@@ -195,41 +209,64 @@ void SpinDensityNew::collect(const RefVector<OperatorEstBase>& type_erased_opera
 {
   auto& dp_ = derived_parameters_;
-  app_log() << pad << "SpinDensity report" << std::endl;
-  app_log() << pad << "  dim     = " << QMCT::DIM << std::endl;
-  app_log() << pad << "  npoints = " << dp_.npoints << std::endl;
-  app_log() << pad << "  grid    = " << dp_.grid << std::endl;
-  app_log() << pad << "  gdims   = " << dp_.gdims << std::endl;
-  app_log() << pad << "  corner  = " << dp_.corner << std::endl;
-  app_log() << pad << "  center  = " << dp_.corner + lattice_.Center << std::endl;
-  app_log() << pad << "  cell " << std::endl;
+  out << pad << "SpinDensity report" << std::endl;
+  out << pad << "  dim     = " << QMCT::DIM << std::endl;
+  out << pad << "  npoints = " << dp_.npoints << std::endl;
+  out << pad << "  grid    = " << dp_.grid << std::endl;
+  out << pad << "  gdims   = " << dp_.gdims << std::endl;
+  out << pad << "  corner  = " << dp_.corner << std::endl;
+  out << pad << "  center  = " << dp_.corner + lattice_.Center << std::endl;
+  out << pad << "  cell " << std::endl;
   for (int d = 0; d < QMCT::DIM; ++d)
-    app_log() << pad << "    " << d << " " << lattice_.Rv[d] << std::endl;
-  app_log() << pad << "  end cell " << std::endl;
-  app_log() << pad << "  nspecies = " << species_.size() << std::endl;
+    out << pad << "    " << d << " " << lattice_.Rv[d] << std::endl;
+  out << pad << "  end cell " << std::endl;
+  out << pad << "  nspecies = " << species_.size() << std::endl;
   for (int s = 0; s < species_.size(); ++s)
-    app_log() << pad << "    species[" << s << "]"
-              << " = " << species_.speciesName[s] << " " << species_size_[s] << std::endl;
-  app_log() << pad << "end SpinDensity report" << std::endl;
+    out << pad << "    species[" << s << "]"
+        << " = " << species_.speciesName[s] << " " << species_size_[s] << std::endl;
+  out << pad << "end SpinDensity report" << std::endl;
 }
 
-void SpinDensityNew::registerOperatorEstimator(hdf_archive& file)
+
+void SpinDensityNew::registerOperatorEstimator(hdf_archive& file) {}
+
+void SpinDensityNew::write(hdf_archive& file)
 {
-  std::vector<size_t> my_indexes;
+  auto writeFullPrecData = [&](auto& file, auto& fp_data) {
+    std::vector<size_t> my_indexes;
+    std::array<hsize_t, 2> ng{1, derived_parameters_.npoints};
+    hdf_path hdf_name{my_name_};
+    file.push(hdf_name);
+    Vector<QMCT::FullPrecRealType> data;
+    std::size_t offset = 0;
+    for (int s = 0; s < species_.size(); ++s)
+    {
+      data.attachReference(fp_data.data() + offset, derived_parameters_.npoints);
+      hid_t g_id = file.push(species_.speciesName[s], true);
+      file.append(data, "value", append_index_);
+      offset += derived_parameters_.npoints;
+      file.pop();
+    }
+    ++append_index_;
+    file.pop();
+  };
 
-  std::vector<int> ng(1, derived_parameters_.npoints);
+#ifdef MIXED_PRECISION
+  std::vector<QMCT::FullPrecRealType> expanded_data(data_.size(), 0.0);
+  std::copy_n(data_.begin(), data_.size(), expanded_data.begin());
```

Comment:
Can we do the cast during construction of `expanded_data`?
```suggestion
  std::vector<QMCT::FullPrecRealType> expanded_data(data_.cbegin(), data_.cend());
```